crates/wasmparser/src/validator/operators.rs

```diff
@@ -1155,6 +1144,40 @@ where
         Ok(ty)
     }
 
+    /// Common helper for checking the types of structs accessed with atomic RMW
+    /// instructions, which only allow `i32` and `i64`.
+    fn check_atomic_struct_rmw_ty(
+        &self,
+        struct_type_index: u32,
+        field_index: u32,
+    ) -> Result<ValType> {
+        let ty = self
+            .struct_field_at(struct_type_index, field_index)?
+            .element_type;
+        match ty {
+            StorageType::Val(ValType::I32) => Ok(ValType::I32),
+            StorageType::Val(ValType::I64) => Ok(ValType::I64),
+            _ => bail!(
+                self.offset,
+                "invalid type: `struct.atomic.rmw.*` only allows `i32` and `i64`"
+            ),
+        }
+    }
+
+    /// Common helper for checking the types of arrays accessed with atomic RMW
+    /// instructions, which only allow `i32` and `i64`.
+    fn check_atomic_array_rmw_ty(&self, type_index: u32) -> Result<ValType> {
+        let elem_ty = self.array_type_at(type_index)?.0.element_type;
+        match elem_ty {
+            StorageType::Val(ValType::I32) => Ok(ValType::I32),
+            StorageType::Val(ValType::I64) => Ok(ValType::I64),
+            _ => bail!(
+                self.offset,
+                "invalid type: `struct.atomic.rmw.*` only allows `i32` and `i64`"
```

Comment:
copy/paste with `struct.atomic.*` here, but perhaps this and the above method could share more implementation details?

Reply:
Fixed the typo and looked at merging them: the only part that is the same is the i32/i64 check. The other parts--type access and push/pop logic--ends up being different.